amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/jsonschema/AstIndex.scala

```diff
@@ -1,18 +1,23 @@
 package amf.plugins.document.webapi.parser.spec.jsonschema
 
-import java.net.URI
-
 import amf.plugins.document.webapi.parser.spec.common.YMapEntryLike
-
+import java.net.URI
 import scala.collection.mutable
 
-case class AstIndex(private val map: mutable.Map[String, YMapEntryLike]) {
+case class AstIndex(private val map: mutable.Map[String, YMapEntryLike], resolvers: Seq[ReferenceResolver]) {
 
   def getNode(reference: String): Option[YMapEntryLike] = {
     val toLookUp = clean(reference)
-    map.get(toLookUp)
+    map.get(toLookUp).orElse { callResolvers(toLookUp) }
   }
 
+  private def callResolvers(reference: String): Option[YMapEntryLike] =
+    resolvers.iterator.map(_.resolve(reference, map.toMap)).collectFirst {
+      case Some(entry) =>
+        map.put(reference, entry)
```

Comment:
isnt this `map` reference an immutable Map?

Reply:
Maybe we should add a separate unit test for this "lazy building"

Reply:
Agree, added specific test for this case.